lib/active_merchant/billing/gateways/versa_pay.rb

```diff
@@ -0,0 +1,307 @@
+        }.compact
+
+        add_invoice(order, money, options)
+        add_address(order, options, 'shipping')
+        add_address(order, options)
+        add_attributes(order, options[:attributes])
+        add_order_items(order, options)
+        post[:order] = order
+      end
+
+      def add_attributes(hash, attributes)
+        return unless attributes
+
+        attributes.each_with_index do |atr, i|
+          hash[:"order_att#{1 + i}"] = atr
+        end
+      end
+
+      def add_order_items(post, options)
+        return unless options[:order_items]
+
+        post[:order_items] = options[:order_items].map.with_index(1) do |order_item, i|
+          {
+            number: i,
+            description: order_item[:description],
+            quantity: order_item[:quantity],
+            unit_cost_cents: order_item[:unit_cost_cents],
+            amount_cents: order_item[:amount_cents],
+            tax_amount_cents: order_item[:tax_amount_cents],
+            discount_amount_cents: order_item[:discount_amount_cents],
+            product_code: order_item[:product_code],
+            category: order_item[:category]
+          }.compact.tap do |order_it|
+            add_attributes(order_it, order_item[:attributes])
+          end
+        end
+      end
+
+      def add_address(post, options, address_key = 'billing', hash = 'order')
+        address = options["#{address_key}_address".to_sym]
+        return unless address
+
+        address_data = {
+          address_1: address[:address1],
+          city: address[:city],
+          province: address[:state],
+          postal_code: address[:zip],
+          country: Country.find(address[:country]).code(:alpha3).value
+        }
+
+        if hash == 'payment_method'
+          post[:address] = address_data
+        else
+          post.merge!(
+            "#{address_key}_name": address[:company],
+            "#{address_key}_address": address[:address1],
+            "#{address_key}_address2": address[:address2],
+            "#{address_key}_city": address[:city],
+            "#{address_key}_country": address_data[:country],
+            "#{address_key}_email": options[:email],
+            "#{address_key}_telephone": address[:phone] || address[:phone_number],
+            "#{address_key}_postalcode": address[:zip],
+            "#{address_key}_state_province": address[:state]
+          ).compact!
+        end
+      end
+
+      def add_payment(post, payment, options)
+        if payment.is_a?(CreditCard)
+          post[:credit_card] = {
+            name: payment.name,
+            expiry_month: format(payment.month, :two_digits),
+            expiry_year: payment.year,
+            card_number: payment.number,
+            cvv: payment.verification_value
+          }
+          add_address(post[:credit_card], options, 'billing', 'payment_method')
+        end
+      end
+
+      def parse(body)
```

Comment:
💬 Comment/Style:

This method is doing to much, maybe some extraction and separation of concerns could clean it a little bit, something like:

```ruby
def parse(body)
  JSON.parse(body).with_indifferent_access.tap do |json|
    parse_inner_json(json)
  end
rescue JSON::ParserError => e
  handle_json_decode_error(body, e)
end

def parse_inner_json(json)
  json.each do |key, _value|
    next unless key.is_a?(String) && key.strip.start_with?('{')

    begin
      parsed_key = JSON.parse(key).with_indifferent_access
      json.merge!(parsed_key)
      json.delete(key)
    rescue JSON::ParserError
      next
    end
  end
end

def handle_json_decode_error(body, e)
  {
    errors: body,
    status: 'Unable to parse JSON response',
    message: e.message
  }.with_indifferent_access
end

```

Reply:
Curiosity this change was required because I used the order field, due to removing that field, I can discard this method.